Drivers/Src/Minions.c

```
@@ -1,406 +1,51 @@
-}
-
-/**
-* @brief   Read the state of a specific light from the light bitmap
-* @param   light Which Light to read
-* @return  returns State enum which indicates ON/OFF
-*/ 
-State Lights_Read(light_t light) {
-    //Return from a stored state bitmap instead of actually querying hardware. 
-    return (State) ((lightStatesBitmap>>light)&0x0001);
-}
-
-/**
-* @brief   Read the lights bitmap
-* @return  returns uint16_t with lights bitmap
-*/ 
-uint16_t Lights_Bitmap_Read() {
-    //Return from a stored state array instead of actually querying hardware.
-    return lightStatesBitmap;
-}
-
-/**
- * @brief   Reads from static variable bitmap holding values of switches
- * @param   sw
- * @return  ON/OFF State
- */ 
-State Switches_Read(switches_t sw){
-    if(sw != LEFT_SW && sw != RIGHT_SW && sw != FOR_SW && sw != REV_SW){
-        return (State) ((switchStatesBitmap >> sw) & 0x0001);
-    }else{
-        if(sw == LEFT_SW || sw == RIGHT_SW){
-            // If Left and Right are one, then in reality, none are selected
-            if(((switchStatesBitmap&0x20) == 0x20)&&((switchStatesBitmap&0x40) == 0x40)){
-                return OFF;
-            }
-            return (State) ((switchStatesBitmap >> sw) & 0x0001);
-        }else{
-            // If Forward and Reverse are one, then in reality, none are selected
-            if(((switchStatesBitmap&0x04) == 0x04)&&((switchStatesBitmap&0x08) == 0x08)){
-                return OFF;
-            }
-            return (State) ((switchStatesBitmap >> sw) & 0x0001);
-        }
+    for(uint8_t i = 0; i < MINIONPIN_NUM; i++){
+        BSP_GPIO_Init(PINS_LOOKARR[i].port, PINS_LOOKARR[i].pinMask, PINS_LOOKARR[i].direction);
     }
-    
+    OSMutexCreate(&OutputMutex, "Minions Output Mutex", &err);
+    assertOSError(OS_MINIONS_LOC, err);
 }
 
+bool Minion_Read_Input(MinionPin_t pin){
+    if((PINS_LOOKARR[pin].direction == OUTPUT)){ //trying to read from an output pin, can't do that.
+        return false; 
+    }   
```

If the direction is wrong, return an error. This requires the creation of an error enum (see [here](https://github.com/lhr-solar/Controls/blob/a4242873c8441336c8aa7f9360020b5b29eadb7c/Drivers/Inc/Display.h#L25-L37)) and an assertion function. Return the error through a pointer parameter and dereference the pointer.

The reasoning behind this is the end user needs to know when a write has failed (also makes it easier to debug)

Actually, we're able to read an output pin, so you may want to return an error here into the parameter (MINION_READ_OUTPUT_ERR or smth) but **_still_** read the input and return it. So returning the error will basically be a warning or a notifier to the user saying "hey you're reading an output but it's not gonna break the car and we're gonna return a valid value anyways". The assertion function can have a switch case that doesn't necessarily go to fault state based off of this.